ortools/math_opt/solvers/xpress/g_xpress.cc

```diff
@@ -0,0 +1,263 @@
+  const int n_coefs = static_cast<int>(rowind.size());
+  auto c_rowind = const_cast<int*>(rowind.data());
+  auto c_colind = const_cast<int*>(colind.data());
+  auto c_values = const_cast<double*>(values.data());
+  return ToStatus(
+      XPRSchgmcoef(xpress_model_, n_coefs, c_rowind, c_colind, c_values));
+}
+
+absl::StatusOr<int> Xpress::LpOptimizeAndGetStatus(std::string flags) {
+  RETURN_IF_ERROR(ToStatus(XPRSlpoptimize(xpress_model_, flags.c_str())))
+      << "XPRESS LP solve failed";
+  int xpress_status;
+  RETURN_IF_ERROR(
+      ToStatus(XPRSgetintattrib(xpress_model_, XPRS_LPSTATUS, &xpress_status)))
+      << "Could not get XPRESS status";
+  return xpress_status;
+}
+absl::Status Xpress::PostSolve() {
+  return ToStatus(XPRSpostsolve(xpress_model_));
+}
+
+absl::StatusOr<int> Xpress::MipOptimizeAndGetStatus() {
+  RETURN_IF_ERROR(ToStatus(XPRSmipoptimize(xpress_model_, nullptr)))
+      << "XPRESS MIP solve failed";
+  int xpress_status;
+  RETURN_IF_ERROR(
+      ToStatus(XPRSgetintattrib(xpress_model_, XPRS_MIPSTATUS, &xpress_status)))
+      << "Could not get XPRESS status";
+  return xpress_status;
+}
+
+void Xpress::Terminate() { XPRSinterrupt(xpress_model_, XPRS_STOP_USER); };
+
+absl::StatusOr<int> Xpress::GetIntAttr(int attribute) const {
+  int result;
+  RETURN_IF_ERROR(ToStatus(XPRSgetintattrib(xpress_model_, attribute, &result)))
+      << "Error getting Xpress int attribute: " << attribute;
+  return result;
+}
+
+absl::Status Xpress::SetIntAttr(int attribute, int value) {
+  return ToStatus(XPRSsetintcontrol(xpress_model_, attribute, value));
+}
+
+absl::StatusOr<double> Xpress::GetDoubleAttr(int attribute) const {
+  double result;
+  RETURN_IF_ERROR(ToStatus(XPRSgetdblattrib(xpress_model_, attribute, &result)))
+      << "Error getting Xpress double attribute: " << attribute;
+  return result;
+}
+
+absl::StatusOr<std::vector<double>> Xpress::GetPrimalValues() const {
+  int nVars = GetNumberOfVariables();
+  XPRSgetintattrib(xpress_model_, XPRS_COLS, &nVars);
+  std::vector<double> values(nVars);
+  RETURN_IF_ERROR(ToStatus(
+      XPRSgetlpsol(xpress_model_, values.data(), nullptr, nullptr, nullptr)))
+      << "Error getting Xpress LP solution";
+  return values;
+}
+
+int Xpress::GetNumberOfConstraints() const {
+  int n;
+  XPRSgetintattrib(xpress_model_, XPRS_ROWS, &n);
+  return n;
+}
+
+int Xpress::GetNumberOfVariables() const {
+  int n;
+  XPRSgetintattrib(xpress_model_, XPRS_COLS, &n);
+  return n;
+}
+
+absl::StatusOr<std::vector<double>> Xpress::GetConstraintDuals() const {
+  int nCons = GetNumberOfConstraints();
+  double values[nCons];
+  RETURN_IF_ERROR(
+      ToStatus(XPRSgetlpsol(xpress_model_, nullptr, nullptr, values, nullptr)))
+      << "Failed to retrieve LP solution from XPRESS";
+  std::vector<double> result(values, values + nCons);
+  return result;
+}
+absl::StatusOr<std::vector<double>> Xpress::GetReducedCostValues() const {
```

Comment:
[XPRSgetredcosts()](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSgetredcosts.html) is the preferred way to query reduced costs.